plugins/document/document.go

```diff
@@ -0,0 +1,219 @@
+// WHERE pk >= "Customer#" AND pk < "Customer!"
+func GetEndRangeValue(value string) string {
+	strFrontCode := value[:len(value)-1]
+
+	strEndCode := value[len(value)-1:]
+
+	return strFrontCode + string(strEndCode[0]+1)
+}
+
+// Validate the provided query expressions
+func ValidateExpressions(expressions []sdk.QueryExpression) error {
+	if expressions == nil {
+		return errors.New("provide non-nil query expressions")
+	}
+
+	inequalityProperties := make(map[string]string)
+
+	for _, exp := range expressions {
+		if exp.Operand == "" {
+			return fmt.Errorf("provide non-blank query expression operand: %v", exp)
+		}
+
+		if _, found := validOperators[exp.Operator]; !found {
+			return fmt.Errorf("provide valid query expression operator [==, <, >, <=, >=, startsWith]: %v", exp.Operator)
+		}
+		if exp.Value == "" {
+			return fmt.Errorf("provide non-blank query expression value: %v", exp)
+		}
+
+		if exp.Operator != "==" {
+			inequalityProperties[exp.Operand] = exp.Operator
+		}
+	}
+
+	// Firestore inequality compatability check
+	if len(inequalityProperties) > 1 {
+		msg := ""
+		for prop, exp := range inequalityProperties {
+			if msg != "" {
+				msg += ", "
+			}
+			msg += prop + " " + exp
+		}
+		return fmt.Errorf("inequality expressions on multiple properties not supported with Firestore: [ %v ]", msg)
+	}
+
+	// DynamoDB range expression compatability check
+	if err := hasRangeError(expressions); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// QueryExpression sorting support with sort.Interface
+
+type ExpsSort []sdk.QueryExpression
+
+func (exps ExpsSort) Len() int {
+	return len(exps)
+}
+
+// Sort by Operand then Operator then Value
+func (exps ExpsSort) Less(i, j int) bool {
+
+	operandCompare := strings.Compare(exps[i].Operand, exps[j].Operand)
+	if operandCompare == 0 {
+
+		// Reverse operator comparison for to support range expressions
+		operatorCompare := strings.Compare(exps[j].Operator, exps[i].Operator)
+		if operatorCompare == 0 {
+
+			return strings.Compare(exps[i].Value, exps[j].Value) < 0
+
+		} else {
+			return operatorCompare < 0
+		}
+
+	} else {
+		return operandCompare < 0
+	}
+}
+
+func (exps ExpsSort) Swap(i, j int) {
+	exps[i], exps[j] = exps[j], exps[i]
+}
+
+// DynamoDB only supports query range operands: >= AND <=
+// For example: WHERE price >= 20.00 AND price <= 50.0
+func hasRangeError(exps []sdk.QueryExpression) error {
+
+	sortedExps := make([]sdk.QueryExpression, len(exps))
+	copy(sortedExps, exps)
+
+	sort.Sort(ExpsSort(sortedExps))
+
+	for index, exp := range sortedExps {
+		if index < (len(sortedExps) - 1) {
+			nextExp := sortedExps[index+1]
+
+			if exp.Operand == nextExp.Operand &&
+				((exp.Operator == ">" && nextExp.Operator == "<") ||
+					(exp.Operator == ">" && nextExp.Operator == "<=") ||
+					(exp.Operator == ">=" && nextExp.Operator == "<")) {
+
+				return fmt.Errorf("range expression not supported with DynamoDB (use operators >= and <=) : %v", exp)
```

Comment:
As we're accepting this as part of our unified API, is it necessary to mention DynamoDB here? Could be confusing for users deployed to other clouds seeing this error.

Reply:
Yep, makes sense.